envoy/network/connection.h

```diff
@@ -326,6 +336,12 @@ class Connection : public Event::DeferredDeletable,
    *  returned.
    */
   virtual absl::optional<std::chrono::milliseconds> lastRoundTripTime() const PURE;
+
+  /**
+   * Calls `getsockopt` on the underlying socket.
+   */
+  virtual Api::SysCallIntResult getSocketOption(int level, int optname, void* optval,
```

Comment:
Yeah I don't love this either. A few options for thought:
1) At minimum, I would make this a const method, even if it requires const_cast internally to implement, as I agree this is logically a const method.
2) One other option would to actually return some type of generic connection info struct. Internally this would be mapped from the linux TCP_INFO struct, and in all other cases just return an empty absl::optional<>. If you went down this path, you could call the filter `connection_info` and just have it be documented that it currently only works on Linux. I like this option as eventually I could see this getting implemented on Windows and it would avoid a bunch of churn.

Reply:
Another idea that just occurred to me: maybe this should be a transport_socket instead of a network filter. It could use `PassthroughSocket`, wrap another "real" transport socket, and I think it gets all the correct calls (such as connection being closed before it is closed). And it has direct access to the ioHandle() which already exposes `getOption()` which wraps `getsockopt()`. That would avoid any changes to the connection interface.

Reply:
Sure that would work also. As I said I still think some level of abstraction here would not be very difficult and make it easy to support on other OS later, but ultimately that's up to you.

Reply:
I looked at it briefly, looking at the stats supported by windows. They are similar, but different enough that it would be hard to abstract. For instance, linux tracks retransmissions by segments, and windows tracks them by bytes. Mac does both. Linux gives a metric for the minimum RTT over some undefined period of time (some complicated window algorithm). Mac gives the average RTT. Windows gives the minRTT over the lifetime of the connection.

So similar information can be derived, but I'm worried that it wouldn't be close enough to call them the same stat name, because a time series containing multiple platforms wouldn't make sense then. For all of these metrics, they're being measured in an OS-dependent way, so I'm worried that trying to make this cross-platform just ends up with weirdness in the platform differences. 

On the other hand, they might be close enough to be useful in the same way. The metrics are at least advertised to be the same in some cases.

Either way, I think doing a transport socket causes less ugliness on the Connection interface.

Reply:
If we want to make it something abstracted, should it include quic? All the same metrics roughly apply to quic as well logically (though I haven't looked at what metrics quiche gives us access to).

Reply:
Yes I think it would be optimal to include QUIC. I guess my advise to abstract only as much as you need for your current case and we can go from there?